feat: Add table component to editor

vsgoulart opened the pull request:
<!--

Thanks for creating this pull request!

Please make sure to link the issue you are closing as "Closes #issueNr". 
This helps us to understand the context of this PR.

-->

Closes #891

- [ ] This PR adds a new `form-js` element or visually changes an existing component.
  * => In that case, we need to ensure we follow up on this, e.g. by [creating an issue in Tasklist](https://github.com/camunda/tasklist/issues/new/choose)

vsgoulart commented on packages/form-js-playground/src/components/PlaygroundRoot.js:
```diff
@@ -137,6 +137,27 @@ export function PlaygroundRoot(props) {
     paletteRef.current = formEditor.get('palette');
     propertiesPanelRef.current = formEditor.get('propertiesPanel');
 
+    formEditor.on('formField.add', ({ formField }) => {
+      const { initialDemoData, id } = formField;
+
+      if ([ initialDemoData, id ].includes(undefined)) {
+        return;
+      }
+
+      setData((currentData) => {
+        const newData = {
+          ...currentData,
+          [id]: initialDemoData,
+        };
+
+        dataEditorRef.current.setValue(
+          toString(newData)
+        );
+
+        return newData;
+      });
+    });
```
Couldn't really find a good way to test this
I'm open to ideas

vsgoulart commented on packages/form-js-playground/src/components/PlaygroundRoot.js:
```diff
@@ -137,6 +137,27 @@ export function PlaygroundRoot(props) {
     propertiesPanelRef.current = formEditor.get('propertiesPanel');
 
+    formEditor.on('formField.add', ({ formField }) => {
+      const { initialDemoData, id } = formField;
```
Yeah, your way is better

pinussilvestrus replied:
What could also be an alternative, to keep these things out of `form-js-viewer`, to have PlaygroundFormFields that override the config in `form-js-playground` similar to the [EditorFormFields](https://github.com/bpmn-io/form-js/blob/develop/packages/form-js-editor/src/render/EditorFormFields.js). This way we could scope these demo data to the Playground only.

Just an idea, I see also good arguments to define these things in the viewer package.

vsgoulart replied:
Even though I guess this is an editor feature I think it's ok to "break" the boundaries a bit for the sake of having all the config in a single place. So I think I'll keep as this if you don't see it as a blocker

vsgoulart commented:
@pinussilvestrus I also updated the schema for the table component a bit, I was going to use the `key` property as the data source, but then this clashes with the `keyed` logic. So I introduced the `dataSource` property to the schema

pinussilvestrus commented on packages/form-js-playground/src/components/PlaygroundRoot.js:
```diff
@@ -137,6 +137,30 @@ export function PlaygroundRoot(props) {
     paletteRef.current = formEditor.get('palette');
     propertiesPanelRef.current = formEditor.get('propertiesPanel');
 
+    formEditor.on('formField.add', ({ formField }) => {
```
What I'm now wondering after having a second look, does this work with `undo` / `redo`? I think undo, redo might anyway not work super sophisticated in the Form Playground, so it might be ok for now.

vsgoulart replied:
This basically resets the edit history, I couldn't find a better API to properly append the history

pinussilvestrus replied:
We have the concept of [modeling behaviors](https://github.com/bpmn-io/form-js/tree/develop/packages/form-js-editor/src/features/modeling/behavior) that intercept `commands` as removing, adding form fields, updating IDs etc. That ensures proper handling of the Form Editor's command stack and ensures we properly clean up things on undo/redo.

In this case, we use a simple event listener that runs _after_ the `formField.add` command, so it would not run on undo/redo, e.g. the JSON editor would not properly remove the demo data. I think that's not a problem currently, but something you could think about in the future to implement this as a proper modeling behavior.

pinussilvestrus replied:
cc @Skaiir

vsgoulart replied:
ok, merging this for now then

Skaiir replied:
@vsgoulart 

What's the use case for the demo data? To show something in the editor?

vsgoulart replied:
> @vsgoulart
> 
> What's the use case for the demo data? To show something in the editor?

@Skaiir yes, the goal is to make it easier to get "onboarded" on this new component so he/she doesn't just see an empty table in the preview

pinussilvestrus commented on packages/form-json-schema/test/spec/validation.spec.js:
```diff
@@ -237,6 +237,8 @@ describe('validation', function() {
 
     testForm('rowCount-not-allowed');
 
+    testForm('dataSource-not-allowed');
```
```suggestion

    testForm('dataSource-not-allowed');
```

pinussilvestrus reviewed (approved):
Looks good from my side 👍 Only one minor thing and a comment.
